presto-twitter-functions/src/main/java/com/facebook/presto/twitter/functions/SnowflakeFunctions.java

```diff
@@ -0,0 +1,129 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.twitter.functions;
+
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.function.Description;
+import com.facebook.presto.spi.function.ScalarFunction;
+import com.facebook.presto.spi.function.SqlType;
+import com.facebook.presto.spi.type.StandardTypes;
+
+import java.util.concurrent.TimeUnit;
+
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+
+/* size in
+ * bits:  1                     41                      5     5        12
+ *       +-+-----------------------------------------+-----+-----+------------+
+ *       |0|              milliseconds               |clstr|instc|  sequence  |
+ *       |0|              since twepoch              |  id |  id |   number   |
+ *       +-+-----------------------------------------+-----+-----+------------+
+ *        |
+ *        +- Most significant bit
+ */
+public class SnowflakeFunctions
+{
+    private static final long SequenceNumBits = 12L;
+    private static final long MaxSequenceNum = (1L << SequenceNumBits) - 1;
+    private static final long SequenceNumMask = MaxSequenceNum;
+
+    private static final long InstanceIdBits = 5L;
+    private static final long MaxInstanceId = (1L << InstanceIdBits) - 1;
+    private static final long InstanceIdShift = 12L;
+    private static final long InstanceIdMask = MaxInstanceId << InstanceIdShift;
+
+    private static final long ClusterIdBits = 5L;
+    private static final long MaxClusterId = (1L << ClusterIdBits) - 1;
+    private static final long ClusterIdShift = InstanceIdShift + InstanceIdBits;
+    private static final long ClusterIdMask = MaxClusterId << ClusterIdShift;
+
+    private static final long TimestampBits = 41L;
+    private static final long MaxTimestamp = (1L << TimestampBits) - 1;
+    private static final long TimestampShift = ClusterIdShift + ClusterIdBits;
+    private static final long TimestampMask = MaxTimestamp << TimestampShift;
+
+    /* Twepoch is 2010-11-04T01:42:54Z.
+     * Value is in millis since Unix Epoch 1970-01-01T00:00:00Z.
+     */
+    private static final long Twepoch = 1288834974657L;
+    private static final long FirstSnowflakeIdUnixTime = Twepoch + TimeUnit.DAYS.toMillis(1); // 1 day after Twepoch.
+    private static final long FirstSnowflakeId = firstSnowflakeIdFor(FirstSnowflakeIdUnixTime);
+
+    private SnowflakeFunctions()
+    {
+    }
+
+    @ScalarFunction("is_snowflake")
+    @Description("Check if a BIGINT is a Snowflake ID")
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean isSnowflakeId(@SqlType(StandardTypes.BIGINT) long id)
+    {
+        return id >= FirstSnowflakeId;
```

Comment:
Can it overflow? For example, in case `id > (TimestampMask << TimestampShift)`, the MSB will be set to 1, which should be a invalid id.

Reply:
I think overflow will be rejected before reaching the scalar function

```
com.facebook.presto.sql.parser.ParsingException: line 1:1: Invalid numeric literal: 2656055881830522880000000
```

Reply:
can you try negative `BIGINT` ? E.g., `-2^63`. Parser reject it because it is larger than the max value of BIGINT which is `2^63 - 1`.

Reply:
I believe a valid id must be positive based on the definition where the most significant bit marked as 0, so any negative value should be considered as an invalid id?